New ringbuffer proposal (#3848)
microsoft/ebpf-for-windows: docs/RingBuffer.md

      if (!have_data) continue;
    } else { // We got notified of new data.
      have_data = true;
    }
  }
  uint64_t remaining = producer_offset - consumer_offset;

  if (remaining == 0) {
    have_data = false; // Caught up to producer.
    continue;
  } else if (remaining < sizeof(rb_header_t)) {
    // Bad record or consumer offset out of alignment.
    // … log error …
    break;
  }

  // Check header flags first, then read/skip data and update offset.
  rb_header_t header = (rb_header_t)(&rb_data[consumer_offset % rb_size]);
  if (header.locked) { // Next record not ready yet, wait.
    have_data = false;
    continue;
    // Or we could spin/poll on ((rb_header_t*)(&rb_data[consumer_offset % rb_size]))->locked.
  }
  if (!header.discarded) {
    const rb_record_t *record = *(const rb_record_t*)(&rb_data[consumer_offset % rb_size]);
    // Read data from record->data[0 ... record->length-1].
    // … business logic …
  } // Else it was discarded, skip and continue.

  // Update consumer offset (and pad record length to multiple of 8).
  consumer_offset += sizeof(rb_header_t) + (record->length + 7 & ~7);

>   WriteRelease64(cons_offset, consumer_offset);

Author: WriteNoFence64

# New ebpf Ring Buffer Map (proposal)

## Overview

The current ringbuffer uses a pure callback-based approach to reading the ringbuffer.
Linux also supports memory-mapped polling consumers, which can't be directly supported in the current model.

The new API will support 2 consumer types: callbacks and direct access to the mapped producer memory (with poll to wait for data).

Callback consumer:

1. Call `ring_buffer__new` to set up callback.
2. The callback will be invoked for each record written to the ring buffer.

Mapped memory consumer:

1. Call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages.
2. Call `ebpf_ring_buffer_get_wait_handle` to get the wait handle.
3. Directly read records from the producer pages (and update consumer offset as we read).

> 4. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.

Reviewer: Line 6 says Linux supports polling consumers.  What APIs do callers use on Linux to get this behavior?
Why can't those APIs work on Windows?  Why are new `ebpf_` APIs needed?

This document doesn't appear to answer any of these questions.

Author: Updated to discuss the the linux APIs. On linux mmap is used for mapping the memory and epoll used for waiting. There aren't functionally equivalent APIs on Windows, so I also added a windows-specific function to get the wait handle (instead of `ring_buffer__epoll_fd()`, since we don't return an epoll fd).

Reviewer: For `mmap`, bing copilot says it can be done if the driver uses `ZwCreateSection`, and the user process does `MapViewOfFile` on the result.  I didn't ask it about epoll.

But could the new apis be implemented on top of the Linux mmap and epoll?  If so, then they're technically not windows specific, right?

Author: Yes, get_buffer and get_wait_handle could be implemented on top of `mmap` and `ring_buffer_epoll_fd` (with an intermediate translation layer), so they aren't necessarily Windows specific.

On linux, those 2 parts of the ringbuffer API are already linux specific though -- they use the returned values directly with linux syscalls to map the memory and wait on the epoll fd.

Redesigning the libbpf API to be more platform independent is possible, but this proposal keeps the rest of the linux libbpf API as-is and just replaces the 2 linux-specific parts with windows-specific functions, plus adds asyc/sync options since linux is synchronous-only. To avoid a compatibility break for existing Windows customers, the default behavior on Windows is also different from Linux.

That redesign effort seems like a separate task, and when that happens, we could replace/wrap the Windows functions with the generic ones.

---

On creating a translation layer to hide the Windows-specific parts:

ZwCreateSection is limited to memory mappings backed by files or the page file, so it is not directly compatible with the ringbuffer mmap usage on linux, and I don't think there are directly compatible APIs to hide the functionality behind something that looks like mmap.

Similarly, epoll's behavior and design is not quite the same as KEVENT. There are some epoll-like APIs implemented in Windows, and we could implement `epoll_fd` to return one of those wrapped in an fd, but we opted not to do that to avoid hiding the differences between the designs, and for it to be intuitive/obvious which other Windows APIs the ring buffer wait handle works with (which on Linux is already the case because the API is linux-specific).

Use the existing `ring_buffer__new()` to set up automatic callbacks for each record.
Call `ebpf_ring_buffer_get_buffer()` ([New eBPF APIs](#new-ebpf-apis-for-mapped-memory-consumer))
to get direct access to the mapped ringbuffer memory.

```c
struct ring_buffer;

typedef int (*ring_buffer_sample_fn)(void *ctx, void *data, size_t size);

struct ring_buffer_opts {
	size_t sz; /* size of this struct, for forward/backward compatiblity */
  uint64_t flags; /* ring buffer option flags */
};

enum ring_buffer_flags {
  RINGBUF_FLAG_NO_AUTO_CALLBACK = (uint64_t)1 << 0 /* Don't automatically invoke callback for each record */
};

#define ring_buffer_opts__last_field sz

/**
 * @brief Creates a new ring buffer manager.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[in] sample_cb Pointer to ring buffer notification callback function.
 * @param[in] ctx Pointer to sample_cb callback function context.
 * @param[in] opts Ring buffer options.
 *
 * @returns Pointer to ring buffer manager.
 */
struct ring_buffer *
ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, void *ctx,
		 const struct ring_buffer_opts *opts);

/**

>  * @brief poll ringbuf for new data (NOT CURRENTLY SUPPORTED)

Reviewer: Why not currently supported?

Author: The design has been updated to support both polling and asynchronously called consumers.

On Linux there are only polling/synchronous consumers, so Linux doesn't need to differentiate between types of consumers here.

The initial ebpf-for-windows ring buffer support was for asynchronous callbacks only, so to maintain behavior for existing ebpf-for-windows users, the default for the ring buffer manager will be automatic callbacks. The new `RINGBUF_FLAG_NO_AUTO_CALLBACK` flag can be set in the options to get behavior compatible with Linux where the callback is only called in `ring_buffer__poll()` and `ring_buffer__consume()`.

## API Changes

### Changes to ebpf helper functions

```c
/**
 * @brief Output record to ringbuf
 *
 * Note newly added flag values (to specify wakeup options).
 *
 * Wakeup options (flags):
 * - 0 (auto/default): Notify if consumer has caught up.
 * - BPF_RB_FORCE_WAKEUP - Always notify consumer.
 * - BPF_RB_NO_WAKEUP - Never notify consumer.
 *
 */
ebpf_result_t
ebpf_ring_buffer_output(ebpf_ring_buffer_t* ring, uint8_t* data, size_t length, size_t flags)
```

### Updated supported libbpf functions

```c

struct ring_buffer;

typedef int (*ring_buffer_sample_fn)(void *ctx, void *data, size_t size);

struct ring_buffer_opts {
	size_t sz; /* size of this struct, for forward/backward compatiblity */
};

#define ring_buffer_opts__last_field sz

/**
 * @brief Creates a new ring buffer manager.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[in] sample_cb Pointer to ring buffer notification callback function.
 * @param[in] ctx Pointer to sample_cb callback function context.
 * @param[in] opts Ring buffer options.
 *
 * @returns Pointer to ring buffer manager.
 */
struct ring_buffer *

> ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, void *ctx,

Reviewer 2: How do we differentiate between callback vs poll?

Author: Updated the design so callback consumers now use the libbpf ring buffer manager, and direct mapped memory consumers use the new functions.

Reviewer: I don't follow.  How does Linux differentiate?  Why would we want a different answer?

Author: Linux doesn't have asynchronous callbacks, so they don't need differentiation. After you create the ringbuffer manager, then you call poll/consume to have the callback _synchronously_ called on available records.

Windows started with async callbacks, so this proposes a new option flag to keep the current default behavior and also support synchronous callbacks

> 		 const struct ring_buffer_opts *opts);

Author: I could add SAL to the markdown (for clarity), but not in libbpf.h (which matches the linux headers).

> 		 const struct ring_buffer_opts *opts);

Reviewer 3: In that case, I don't think it is necessary.

    uint8_t locked : 1;
    uint8_t discarded : 1;
    uint32_t length : 30;
    uint32_t offset; // for kernel use (offset of record in pages from start of buffer data area)
} rb_header_t;

typedef struct _rb_record
{
    rb_header_t header;
    uint8_t data[];
} rb_record_t;

/**
 * @brief clear the ringbuffer.
 */
void rb_flush(uint64_t *cons_offset, const uint64_t *prod_offset) {
    *cons_offset = *prod_offset;
}

//
// == mmap/epoll consumer ==
//

void *rb_cons; // Pointer to read/write mapped consumer page with consumer offset.
void *rb_prod; // Pointer to start of read-only producer pages.

// Open ringbuffer.
fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
if (map_fd == ebpf_fd_invalid) return 1;

auto rb = ring_buffer__new(
    map_fd,
    NULL, //callback function for callback-based consumer
    nullptr, nullptr);
if (!rb) return 1;

// Initialize iocp wait handle.
HANDLE wait_handle = ebpf_ring_buffer_get_wait_handle(rb);
if (!wait_handle) {
    // … log error …
    goto Cleanup;
}

// get pointers to the producer/consumer pages

> int err = ebpf_ring_buffer_get_buffer(&rb_prod, &rb_cons);

Author: Whoops, missed the argument in the example code. Thanks, added.

if (err) {
    goto Cleanup;
}

const uint64_t *prod_offset = (const uint64_t*)rb_prod; // Producer offset ptr (r only).
uint64_t *cons_offset = (uint64_t*)rb_cons; // Consumer offset ptr (r/w mapped).
const uint8_t *rb_data = ((const uint8_t*)rb_prod) + PAGESIZE; // Double-mapped rb data ptr (r only).

// have_data used to track whether we should wait for notification or just keep reading.
bool have_data = *prod_offset > *cons_offset;

void           *lp_ctx = NULL;
OVERLAPPED     *overlapped = NULL;
DWORD          bytesTransferred = 0;

// Now loop until error.
For(;;) {
  if (!have_data) { // Only wait if we have already caught up.
    // Wait for rb to notify -- or we could spin/poll on *prod_offset > *cons_offset.
    DWORD wait_status = WaitForSingleObject(wait_handle, INFINITE);

    if (wait_status != WAIT_OBJECT_0) { // No notification
      uint32_t wait_err = GetLastError();
      if (wait_err == /* terminal error */) {
        // … log error …
        break;
      }

>       have_data = *prod_offset > *cons_offset; // It's possible we still have data.

Author: `*cons_offset` shouldn't change from outside us if we only allow a single consumer. `*prod_offset` is updated by the producer as they allocate entries in the ringbuffer.

The producer will never overrun the consumer (writes to a full ringbuf fail), so we just have to check if the producer offset is past the consumer offset to know if we have more records to read. If the next record isn't ready yet then we wait until it is to continue.

We only update the consumer offset after we are done with a record so if we find an unlocked record that is between the consumer and producer offsets we know it won't change from under us.

>       have_data = *prod_offset > *cons_offset; // It's possible we still have data.

Reviewer 2: You need a ReadAcquire here (or an explicit memory barrier). With out that, you can potentially read the record before its been written to by the producer. Likewise you will need a WriteRelease in the producer.

Reviewer 2: Also, reading it multiple times like this is going to cause issues. It should read prod_offset once using ReadAcquire and then use the value.

Author: Now uses ReadAcquire64/WriteRelease64 for access to the producer and consumer offset.

I went with bit fields for the record header for simpler example code, so that part is already pseudocode and I didn't add the ReadAcquire needed to read the header, but added a comment in the struct.

 *
 */
void ring_buffer__free(struct ring_buffer *rb);

/**
 * @brief poll ringbuf for new data
 * Poll for available data and consume records, if any are available.
 * Returns number of records consumed (or INT_MAX, whichever is less), or
 * negative number, if any of the registered callbacks returned error.
 *
 * @param[in] rb Pointer to ring buffer manager.
 * @param[in] timeout_ms maximum time to wait for (in milliseconds).
 *
 */
int ring_buffer__poll(struct ring_buffer *rb, int timeout_ms);

/**
 * @brief catch consumer up to producer by invoking the callback for every available record
 * Consume available ring buffer(s) data without event polling.
 * Returns number of records consumed across all registered ring buffers (or
 * INT_MAX, whichever is less), or negative number if any of the callbacks
 * return error.
 *
 * @param[in] rb Pointer to ring buffer manager.
 */
int ring_buffer__consume(struct ring_buffer *rb);
```

### New ebpf APIs

```c
/**
 * get pointers to mapped producer and consumer pages
 *
 * @param[out] producer pointer* to start of read-only mapped producer pages
 * @param[out] consumer pointer* to start of read-write mapped consumer page
 */

> int ebpf_ring_buffer_get_buffer(const ebpf_map_t* map, void **producer, void **consumer);

Author: To resolve #3914 I'll be updating this to return a handle the user can use to release the mapped memory. This will probably be by using a section object to map views of the producer and consumer pages into memory.

Author: If we map views of sections into user space the user process can unmap the sections using the producer/consumer addresses, so the existing signature supports that.

The new API will support 2 access methods: callbacks and mmap/epoll style access.

callback consumer:

1. call `ring_buffer__new` to set up callback
2. call `ring_buffer__consume` as needed to invoke the callback on any outsanding data that is ready

mmap/epoll consumer:

1. call `ring_buffer__new` to get a ringbuffer manager
2. call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages
3. directly read records from the producer pages (and update consumer offset as we read)
4. call `ring_buffer__epoll_fd` to get the iocp handle as a fd_t so we can wait for new data
5. use `GetQueuedCompletionStatus`/`GetQueuedCompletionStatusEx` to wait for new data to be available
    - We could provide a wrapper to make the common ebpf use cases simpler

## API Changes

### Changes to ebpf helper functions

```c
/**
 * @brief Output record to ringbuf
 *
 * Note newly added flags (to specify wakeup options).
 *
 * Wakeup options:
 * - auto (default): Notify if consumer has caught up.
 * - BPF_RB_FORCE_WAKEUP - Always notify consumer.
 * - BPF_RB_NO_WAKEUP - Never notify consumer.
 *
 */
ebpf_result_t

> ebpf_ring_buffer_output(ebpf_ring_buffer_t* ring, uint8_t* data, size_t length, size_t flags)

Reviewer 4: just to confirm: is this changing the helper function signature for `bpf_ringbuf_output()`?

Author: Actually no. I don't know what I was looking at before thinking we didn't have flags, but we do so there is no change to the signature (I'll edit the md).

```

> ### Updated libbpf ringbuf API

Reviewer 4: nit: to confirm: does this mean that we will be implementing these new libbpf APIs in Windows or does it mean we are modifying the API signature of the libbpf APIs?

Author: Just implementing new libbpf apis (with the same signature as on linux), plus the functions we already have on windows (like output).

Author: The one possible exception to having the same signature is epoll_fd (see notes in markdown).

} rb_record_t;

/**
 * @brief clear the ringbuffer.
 */
void rb_flush(uint64_t *cons_offset, const uint64_t *prod_offset) {
    *cons_offset = *prod_offset;
}

//
// == mmap/epoll consumer ==
//

void *rb_cons; // Pointer to read/write mapped consumer page with consumer offset.
void *rb_prod; // Pointer to start of read-only producer pages.

// Open ringbuffer.
fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
if (map_fd == ebpf_fd_invalid) return 1;

auto rb = ring_buffer__new(
    map_fd,
    NULL, //callback function for callback-based consumer
    nullptr, nullptr);
if (!rb) return 1;

// get pointers to the producer/consumer pages
int err = ebpf_ring_buffer_get_buffer(&rb_prod, &rb_cons, /* &epoll_fd */);

if (err) {
    goto Cleanup;
}

const uint64_t *prod_offset = (const uint64_t*)rb_prod; // Producer offset ptr (r only).
uint64_t *cons_offset = (uint64_t*)rb_cons; // Consumer offset ptr (r/w mapped).
const uint8_t *rb_data = ((const uint8_t*)rb_prod) + PAGESIZE; // Double-mapped rb data ptr (r only).

// have_data used to track whether we should wait for notification or just keep reading.
bool have_data = *prod_offset > *cons_offset;

// Initialize iocp wait handle.

> epoll_fd = ring_buffer__epoll_fd(rb);

Reviewer 5: Should this come before line 193?

Author: It doesn't have to but that probably makes more sense. The code (that was) on line 194 gets points to the mapped memory, and could be an alternative way to get a waitable handle (instead of using epoll_fd to get an integer that can be converted to a HANDLE).

I am modifying this based on discussions, so the next version won't be calling `epoll_fd()` at all (just `poll()`), so we can match the libbpf API without needing extra wrappers for epoll.

# New ebpf Ring Buffer Map (proposal)

## Overview

The current ringbuffer uses a pure callback-based approach to reading the ringbuffer.

> Linux also supports memory-mapped polling consumers, which can't be directly supported in the current model.

Reviewer: What happens if you try to port code to do so?  Would it fail to compile, compile and crash when run, compile and have some other behavior when run, or what?

Author: The APIs to support a memory mapped consumer are not currently exposed, so you wouldn't be able to compile it.

The current implementation uses locking to serialize producers and consumers and doesn't expose the producer page in the memory mapping, so you couldn't have a linux-style consumer that directly reads/writes the producer/consumer offsets and can tell when the next record is ready.

Callback consumer:

1. Call `ring_buffer__new` to set up callback.
2. The callback will be invoked for each record written to the ring buffer.

Mapped memory consumer:

1. Call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages.
2. Call `ebpf_ring_buffer_get_wait_handle` to get the wait handle.
3. Directly read records from the producer pages (and update consumer offset as we read).
4. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.

### Differences from linux API

Linux has only polling ring buffer consumers, even when callbacks are used.
On linux the user code can call `ring_buffer__consume()` to invoke the callback on all available data,
and `ring_buffer__poll()` to wait for data if needed and then consume available data.
Linux consumers can also directly read from the mapped memory by using `mmap()` to map the data
into user space and `ring_buffer__epoll_fd()` to get an epoll wait handle.

On Windows asynchronous events are supported by default,
so nothing extra needs to be done for callbacks to be invoked.

> If the `RINGBUF_FLAG_NO_AUTO_CALLBACK` flag is set, callbacks will not automatically be called and `ring_buffer__poll()`

Reviewer: Why should the default be different between Windows vs Linux?
Why not a `RINGBUF_FLAG_AUTO_CALLBACK` flag instead?

Author: To maintain current default behavior for existing ebpf-for-windows customers.

Author: After discussion, this will be a major breaking change, but we'll want to eventually make it to have the same default behavior across platforms, so this is tracked by #4142.

should be called to poll for available data and invoke the callback. On Windows a timeout of zero can be passed to
`ring_buffer__poll()` to get the same behaviour as `ring_buffer__consume()`.

For direct memory mapped consumers on Windows, use `ebpf_ring_buffer_get_buffer` to get pointers to the producer and consumer
pages mapped into user space, and `ebpf_ring_buffer_get_wait_handle()` to get the SynchronizationEvent (auto-reset) KEVENT
to use with `WaitForSingleObject`/`WaitForMultipleObject`.

Similar to the linux memory layout, the first page of the producer and consumer memory is the "producer page" and "consumer page",
which contain the 64 bit producer and consumer offsets as the first 8 bytes.
Only the producer may update the producer offset, and only the consumer may update the consumer offset.

## ebpf-for-windows API Changes

### Changes to ebpf helper functions

```c
/**
 * @brief Output record to ringbuf
 *
 * Note newly added flag values (to specify wakeup options).
 *
 * Wakeup options (flags):
 * - 0 (auto/default): Notify if consumer has caught up.

>  * - BPF_RB_FORCE_WAKEUP - Always notify consumer.

Reviewer: nit: Inconsistent use of punctuation (`:` vs `-`)

Author: fixed

 * - BPF_RB_NO_WAKEUP - Never notify consumer.
 *
 */
ebpf_result_t

> ebpf_ring_buffer_output(ebpf_ring_buffer_t* ring, uint8_t* data, size_t length, size_t flags)

Reviewer: Please add SAL annotations to ebpf_* APIs in this doc.

Author: I added them to all the functions (for clarity), though they'll only be in the ebpf_* functions in the implementation of this proposal.

```

### Updated libbpf API for callback consumer

The behaviour of these functions will be unchanged.

Use the existing `ring_buffer__new()` to set up automatic callbacks for each record.
Call `ebpf_ring_buffer_get_buffer()` ([New eBPF APIs](#new-ebpf-apis-for-mapped-memory-consumer))
to get direct access to the mapped ringbuffer memory.

```c
struct ring_buffer;

typedef int (*ring_buffer_sample_fn)(void *ctx, void *data, size_t size);

struct ring_buffer_opts {
	size_t sz; /* size of this struct, for forward/backward compatiblity */

>   uint64_t flags; /* ring buffer option flags */

Reviewer: Fix indentation so these lines match

Author: fixed

};

enum ring_buffer_flags {
  RINGBUF_FLAG_NO_AUTO_CALLBACK = (uint64_t)1 << 0 /* Don't automatically invoke callback for each record */
};

#define ring_buffer_opts__last_field sz

/**
 * @brief Creates a new ring buffer manager.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[in] sample_cb Pointer to ring buffer notification callback function.
 * @param[in] ctx Pointer to sample_cb callback function context.
 * @param[in] opts Ring buffer options.
 *
 * @returns Pointer to ring buffer manager.
 */
struct ring_buffer *
ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, void *ctx,
		 const struct ring_buffer_opts *opts);

/**
 * @brief poll ringbuf for new data
 * Poll for available data and consume records, if any are available.
 * Returns number of records consumed (or INT_MAX, whichever is less), or

>  * negative number, if any of the registered callbacks returned error.

Reviewer: Remove, this is redundant with line 117.  (Any elaboration should be added to line 117)

 *
 * If timeout_ms is zero, poll will not wait but only invoke the callback on records that are ready.
 * If timeout_ms is -1, poll will wait until data is ready (no timeout).
 *
 * This function is only supported when the RINGBUF_FLAG_NO_AUTO_CALLBACK flag is set.
 *
 * @param[in] rb Pointer to ring buffer manager.
 * @param[in] timeout_ms maximum time to wait for (in milliseconds).
 *
 * @returns number of records consumed, INT_MAX, or a negative number on error
 */
int ring_buffer__poll(struct ring_buffer *rb, int timeout_ms);

/**
 * @brief Frees a ring buffer manager.
 *
 * @param[in] rb Pointer to ring buffer manager to be freed.
 */
void ring_buffer__free(struct ring_buffer *rb);
```

### New ebpf APIs for mapped memory consumer

```c
/**

>  * get pointers to mapped producer and consumer pages

Reviewer: ```suggestion
 * Get pointers to mapped producer and consumer pages.
```

 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[out] producer pointer* to start of read-only mapped producer pages
 * @param[out] consumer pointer* to start of read-write mapped consumer page
 *

>  * @returns EBPF_SUCCESS on success, or error

Reviewer: Use retval rather than returns, e.g.
```suggestion
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
```

Author: fixed

 */
ebpf_result_t ebpf_ring_buffer_get_buffer(fd_t map_fd, void **producer, void **consumer);

/**

>  * get the wait handle to use with WaitForSingleObject/WaitForMultipleObject

Reviewer: ```suggestion
 * Get the wait handle to use with WaitForSingleObject/WaitForMultipleObject.
```

Author: fixed

 *
 * @param[in] map_fd File descriptor to ring buffer map.
 *
 * @returns Wait handle
 */
HANDLE ebpf_ring_buffer_get_wait_handle(fd_t map_fd);
```

### New user-space helpers for memory mapped consumer

```c
/**
 * The below helpers simplify memory-mapped consumer logic
 * by abstracting operations on the producer and consumer offsets.
 */

/**
 * Get pointer to consumer offset from consumer page.
 *
 * @param[in] cons pointer* to start of read-write mapped consumer page
 *
 * @returns Pointer to consumer offset
 */

> uint64_t* rb__consumer_offset(void *cons);

Reviewer: Does this exact same prototype exist on Linux?  I couldn't find it, I could only find
https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L4558 as a helper function.
This doesn't follow https://github.com/microsoft/ebpf-for-windows/blob/main/docs/DevelopmentGuide.md#coding-conventions so is only ok if it's identical to Linux.

Author: I've removed that set of helpers -- replaced by struct definitions for the ebpf-for-windows producer and consumer pages.

/**
 * Get pointer to producer offset from producer page.
 *
 * @param[in] prod pointer* to start of read-only mapped producer pages
 *
 * @returns Pointer to producer offset
 */

> volatile const uint64_t* rb__producer_offset(volatile const void *prod);

Reviewer: Does this exact same prototype exist on Linux? I couldn't find it, I could only find
https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L4558

Author: Has been removed.

if (!wait_handle) {
    // … log error …
    goto Exit;
}

uint32_t wait_err = 0;

// Consumer loop.
for(;;) {
  for(; !(err=rb__get_record(prod,cons,&record)); rb__next_record(prod,cons)) {
    // Data is now in record->data[0 ... record->length-1].
    // … Do record handling here …
  }
  // 3 cases for err:
  // 1) Ringbuf empty - Wait on handle, or poll for !rb__empty(prod,cons).
  // 2) Record locked - Wait on handle, or spin/poll on header lock bit.
  // 3) Corrupt record or consumer offset - Break (could flush to continue reading from next good record).
  if (err!=E_EMPTY && err!=E_LOCKED) {
    // … log error …
    break;
  }
  DWORD wait_status = WaitForSingleObject(wait_handle, INFINITE);

  if (wait_status != WAIT_OBJECT_0) { // No notification
    wait_err = GetLastError();
    if (wait_err == /* terminal error */) {
      // … log error …
      break;
    }
  }
}

```

### Polling ring buffer consumer (linux-style)

```c
// sample callback
int ring_buffer_sample_fn(void *ctx, void *data, size_t size) {
  // … business logic to handle record …
}

// consumer code
struct ring_buffer_opts opts;
opts.sz = sizeof(opts);

> opts.flags = RINGBUF_FLAG_NO_AUTO_CALLBACK; //no automatic callbacks

Reviewer: One should not need to opt into Linux behavior, it should be the default.
We want to avoid gratuitous differences.

Author: I agree it is not optimal, but the reason for this is avoid a compatibility break for existing ebpf-for-windows customers.

On Windows a timeout of zero can be passed to `ring_buffer__poll()` to get the same behaviour as `ring_buffer__consume()` (consume available records without waiting).

When #4142 is resolved the default behaviour will be changed from asynchronous (automatic) to synchronous callbacks,
so `RINGBUF_FLAG_AUTO_CALLBACK` should always be specified for asynchronous callbacks for forward-compatibility.

#### Memory mapped consumers

As an alternative to callbacks, Linux ring buffer consumers can directly access the
ring buffer data by calling `mmap()` on a ring_buffer map fd to map the data into user space.
`ring_buffer__epoll_fd()` is used (on Linux) to get an fd to use with epoll to wait for data.

Windows doesn't have identical or directly compatible APIs to Linux mmap and epoll, so instead we will perfom the mapping
in the eBPF core and use KEVENTs to signal for new data.

For direct memory mapped consumers on Windows, use `ebpf_ring_buffer_get_buffer` to get pointers to the producer and consumer
pages mapped into user space, and `ebpf_ring_buffer_get_wait_handle()` to get the SynchronizationEvent (auto-reset) KEVENT
to use with `WaitForSingleObject`/`WaitForMultipleObject`.

> Similar to the linux memory layout, the first page of the producer and consumer memory is the "producer page" and "consumer page",

Reviewer 6: Are you going to document the layout, or are users required to use the C helpers to retrieve offsets?

Author: We will document at least the part of the layout that is part of the producer-consumer interface.

The updated document has a struct for the consumer view of the producer and consumer pages.

 *
 * If timeout_ms is zero, poll will not wait but only invoke the callback on records that are ready.
 * If timeout_ms is -1, poll will wait until data is ready (no timeout).
 *
 * This function is only supported when automatic callbacks are disabled (see RINGBUF_FLAG_NO_AUTO_CALLBACK).
 *
 * @param[in] rb Pointer to ring buffer manager.
 * @param[in] timeout_ms maximum time to wait for (in milliseconds).
 *
 * @returns number of records consumed, INT_MAX, or a negative number on error
 */
int ring_buffer__poll(_In_ struct ring_buffer *rb, int timeout_ms);

/**
 * @brief Frees a ring buffer manager.
 *
 * @param[in] rb Pointer to ring buffer manager to be freed.
 */
void ring_buffer__free(_Frees_ptr_opt_ struct ring_buffer *rb);
```

### New ebpf APIs for mapped memory consumer

```c
/**
 * Get pointers to mapped producer and consumer pages.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[out] producer pointer* to start of read-only mapped producer pages
 * @param[out] consumer pointer* to start of read-write mapped consumer page
 *
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
 */

> ebpf_result_t ebpf_ring_buffer_get_buffer(fd_t map_fd, _Out_ void **producer, _Out_ void **consumer);

Reviewer 6: Does this return the same buffer every time it's called? How long is the mapping valid for? Until the fd is closed?

Author: I've updated the doc - the function that takes an `fd_t` creates a new mapping into user-space. The function that takes a ring buffer manager returns the existing mapping that was created in `ring_buffer__new`.

/**
 * Get the wait handle to use with WaitForSingleObject/WaitForMultipleObject.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 *
 * @returns Wait handle
 */

> HANDLE ebpf_ring_buffer_get_wait_handle(fd_t map_fd);

Reviewer 6: ```suggestion
ebpf_handle_t ebpf_ring_buffer_get_wait_handle(fd_t map_fd);
```

Reviewer 6: Same question as above, can I get multiple handles? Do I need to close this?

Author: This has been changed - the handle now comes from user-mode, so the lower level ebpf-for-windows function taking an `fd_t` is to set the handle, and the `get_wait_handle` function on the ring buffer manager returns the wait handle that was created in `ring_buffer__new`.

The user should close the handle as normal unless using the ring-buffer-manager provided handle; the ringbuf manager handle will be closed in `ring_buffer__free`.

 */
int ebpf_ring_buffer_empty(volatile const void *prod, const void *cons);

/**
 * Clear the ring buffer by flushing all completed and in-progress records.
 *
 * This helper just sets the consumer offset to the producer offset
 *
 * @param[in] prod pointer* to start of read-only mapped producer pages
 * @param[in,out] cons pointer* to start of read-write mapped consumer page
 */
void ebpf_ring_buffer_flush(volatile const void *prod, void *cons);

/**
 * Advance consumer offset to next record (if any)
 *
 * @param[in] prod pointer* to start of read-only mapped producer pages
 * @param[in,out] cons pointer* to start of read-write mapped consumer page
 */
void ebpf_ring_buffer_next_record(volatile const void *prod, void *cons);

/**
 * Get record at current ringbuffer offset.

 * @param[in] prod pointer* to start of read-only mapped producer pages
 * @param[in] cons pointer* to start of read-write mapped consumer page
 *
 * @returns E_SUCCESS (0) if record ready, E_LOCKED if record still locked, E_EMPTY if consumer has caught up.
 */
int ebpf_ring_buffer_get_record(volatile const void *prod, const void *cons, volatile const void** record);

```

## Ringbuffer consumer

> ### mapped memory consumer example

Author: Add Linux consumer code examples for comparison.

Author: Completed the Linux examples.

    fd_t map_fd,
    _Out_ ebpf_ring_buffer_consumer_page_t **consumer,
    _Out_ const ebpf_ring_buffer_producer_page_t **producer,
    _Out_ const uint8_t **data,
    _Out_ const uint64_t *data_size
    );

/**
 * Set the wait handle that will be signaled for new data.
 *
 * @note Overwrites the wait handle currently stored in the map.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 *
 * @returns Wait handle
 */
ebpf_result_t ebpf_ring_buffer_map_set_wait_handle(fd_t map_fd, HANDLE handle);
```

### Ringbuffer consumer

#### mapped memory consumer example

This consumer directly accesses the records from the producer memory and directly updates the consumer offset to show the logic. Normally user code should use the ring buffer helpers
(see second example below) to simplify the logic.

```c++

//
// == Direct mapped memory consumer ==
//

// Open ringbuffer.
fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
if (map_fd == ebpf_fd_invalid) return 1;

// Create wait handle.
HANDLE wait_handle = CreateEvent(nullptr, true, false, nullptr);
if (wait_handle == NULL) {
    // … log error …
    goto Exit;
}

// Set map wait handle.

> ebpr_result result = ebpf_ring_buffer_set_wait_handle(map_fd, wait_handle);

Reviewer 6: Nit: `ebpf_ring_buffer_map_set_wait_handle`.

```

#### Updated libbpf API for callback consumer

The behaviour of these functions will be unchanged for now.

Use the existing `ring_buffer__new()` to set up automatic callbacks for each record.
Call `ebpf_ring_buffer_get_buffer()` ([New eBPF APIs](#new-ebpf-apis-for-mapped-memory-consumer))
to get direct access to the mapped ringbuffer memory.

```c
struct ring_buffer;

typedef int (*ring_buffer_sample_fn)(_Inout_ void *ctx, _In_reads_bytes_(size) void *data, size_t size);

struct ring_buffer_opts {
  size_t sz; /* size of this struct, for forward/backward compatiblity */
  uint64_t flags; /* ring buffer option flags */
};

// Ring buffer manager options.
// - The default behaviour is currently automatic callbacks, but may change in the future per #4142.
// - Only specify one of AUTO_CALLBACKS or NO_AUTO_CALLBACKS - specifying both is not allowed.
enum ring_buffer_flags {
  RINGBUF_FLAG_AUTO_CALLBACK = (uint64_t)1 << 0 /* Automatically invoke callback for each record */
  RINGBUF_FLAG_NO_AUTO_CALLBACK = (uint64_t)2 << 0 /* Don't automatically invoke callback for each record */
};

#define ring_buffer_opts__last_field sz

/**
 * @brief Creates a new ring buffer manager.
 *

>  * @param[in] map_fd File descriptor to ring buffer map.

Reviewer 6: P.S. This API is probably not relevant to us, so feel free to ignore.

>  * @param[in] map_fd File descriptor to ring buffer map.

Author: For consistency with current ebpf-for-windows APIs for now it can take ownership of the map.

I know you don't plan on using this, but will this be an incompatibility moving forward between windows/linux? If so then we could change this in the proposal but will also need a separate issue and matching changes in other places (unless already covered by existing map fd issues).

>  * @param[in] map_fd File descriptor to ring buffer map.

Reviewer 6: I'm not familiar with libbpf APIs, sorry.

>  * @param[in] map_fd File descriptor to ring buffer map.

Author: OK. Then we can stick with the existing behavior.

 * @param[in] sample_cb Pointer to ring buffer notification callback function.
 * @param[in] ctx Pointer to sample_cb callback function context.
 * @param[in] opts Ring buffer options.
 *

>  * @returns Pointer to ring buffer manager.

Author: Added a note -- errors returned in `errno`.

There can only be one attached consumer, so multiple calls will fail.

		 _In_ const struct ring_buffer_opts *opts);

/**
 * @brief poll ringbuf for new data
 * Poll for available data and consume records, if any are available.
 *
 * If timeout_ms is zero, poll will not wait but only invoke the callback on records that are ready.
 * If timeout_ms is -1, poll will wait until data is ready (no timeout).
 *
 * This function is only supported when automatic callbacks are disabled (see RINGBUF_FLAG_NO_AUTO_CALLBACK).
 *
 * @param[in] rb Pointer to ring buffer manager.
 * @param[in] timeout_ms maximum time to wait for (in milliseconds).
 *
 * @returns number of records consumed, INT_MAX, or a negative number on error
 */
int ring_buffer__poll(_In_ struct ring_buffer *rb, int timeout_ms);

/**
 * @brief Frees a ring buffer manager.
 *
 * @param[in] rb Pointer to ring buffer manager to be freed.
 */
void ring_buffer__free(_Frees_ptr_opt_ struct ring_buffer *rb);
```

### New ebpf APIs for mapped memory consumer

```c
/**
 * Get the wait handle to use with WaitForSingleObject/WaitForMultipleObject.
 *
 * Get's the wait handle maintained by the ring buffer manager.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 *
 * @returns Wait handle
 */

> ebpf_handle ebpf_ring_buffer_get_wait_handle(_In_ struct ring_buffer *rb);

Author: Split this in two. At the core level the ring stores a user-mode supplied handle.

The ring buffer manager creates and writes a handle to the map which the user code can then get from the ringbuffer manager.

typedef struct _ebpf_ring_buffer_consumer_page
{

>     volatile size_t consumer_offset; ///< Consumer has read up to this offset.

Author: Thanks, fixed

{
    volatile size_t producer_offset; ///< Producer(s) have reserved up to this offset.
} ebpf_ring_buffer_producer_page_t;

/**
 * Map ring buffer memory into user space and get pointers to the consumer, producer, and data regions.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[out] producer pointer* to start of read-only mapped producer pages
 * @param[out] consumer pointer* to start of read-write mapped consumer page
 *
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
 */
ebpf_result_t ebpf_ring_buffer_get_buffer(
    _In_ struct ring_buffer *rb,
    _Out_ ebpf_ring_buffer_consumer_page_t **consumer,
    _Out_ const ebpf_ring_buffer_producer_page_t **producer,
    _Out_ uint8_t **data
    );

/**
 * Map ring buffer memory into user space and get pointers to the consumer, producer, and data regions.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[out] producer pointer* to start of read-only mapped producer pages
 * @param[out] consumer pointer* to start of read-write mapped consumer page
 *
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
 */

> ebpf_result_t ebpf_ring_buffer_map_map_buffer(

Reviewer 6: Linux has user-space writable ring buffers, so you might need some sort of flags argument down the line (or a new function) if you want to support those at some point.

Is this function necessary given that the physical layout of the ring is documented below? What would you lose if we created the mappings ourselves?

Author: There isn't a way for user-space to create the mappings, it needs to be done in the kernel (for the windows APIs that we are using to set up the mapping).

If you mean leave it up to the user to split up the consumer page, producer page, and data I am open to that.

For a ringbuffer specific function it seemed nicer to have typed pointers to reduce void* casting, but we could make this a generic `map_buffer` function that works on any map.

_Note:_ The underlying current ioctl is `map_query_buffer`

Reviewer 6: > If you mean leave it up to the user to split up the consumer page, producer page, and data I am open to that.
> For a ringbuffer specific function it seemed nicer to have typed pointers to reduce void* casting, but we could make this a generic map_buffer function that works on any map.

Yes, that's what I would like to do. Not a deal breaker though, we can invoke the function as outlined since it only needs to happen once on setup.

FWIW Linux allows mapping MAP_TYPE_ARRAY into user space via mmap, and the mapping can be read-only and read-write. This is used to implement efficient sharing of data between user space and BPF programs. So a generic API to map a map might make sense.

Author: I agree. I added a generic `_map_map_buffer` call, with the ringbuffer helper just being a wrapper around it.

    fd_t map_fd,
    _Out_ ebpf_ring_buffer_consumer_page_t **consumer,
    _Out_ const ebpf_ring_buffer_producer_page_t **producer,
    _Out_ const uint8_t **data
    );

/**

>  * Set the wait handle that will be signaled for new data.

Reviewer 6: I'm guessing that multiple calls to this function will overwrite the previously set wait handle? Might be useful to document that as well.

Author: Yes, I'll add notes to document that.

/**
 * Advance consumer offset to next record (if any)
 *
 * @param[in] prod pointer* to start of read-only mapped producer pages
 * @param[in,out] cons pointer* to start of read-write mapped consumer page
 */
void ebpf_ring_buffer_next_record(volatile const void *prod, void *cons);

/**
 * Get record at current ringbuffer offset.

 * @param[in] prod pointer* to start of read-only mapped producer pages
 * @param[in] cons pointer* to start of read-write mapped consumer page
 *
 * @returns E_SUCCESS (0) if record ready, E_LOCKED if record still locked, E_EMPTY if consumer has caught up.
 */
int ebpf_ring_buffer_get_record(volatile const void *prod, const void *cons, volatile const void** record);

```

### Ringbuffer consumer

#### mapped memory consumer example

This consumer directly accesses the records from the producer memory and directly updates the consumer offset to show the logic. Normally user code should use the ring buffer helpers
(see second example below) to simplify the logic.

```c++

//
// == Direct mapped memory consumer ==
//

// Open ringbuffer.
fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
if (map_fd == ebpf_fd_invalid) return 1;

// Create wait handle.

> HANDLE wait_handle = CreateEvent(nullptr, true, false, nullptr);

Reviewer 6: Would it be possible to use an auto resetting event as well?

if (wait_handle == NULL) {
    // … log error …
    goto Exit;
}

// Set map wait handle.
ebpr_result result = ebpf_ring_buffer_set_wait_handle(map_fd, wait_handle);
if (result != EBPF_SUCCESS) {
    // … log error …
    goto Exit;
}

ebpf_ring_buffer_consumer_page_t *rb_cons; // Read/write consumer page.
const ebpf_ring_buffer_producer_page_t *rb_prod; // Read-only producer page.
const uint8_t* data; // Data region for records.

// Get pointers to the 3 regions.
result = ebpf_ring_buffer_map_map_buffer(map_fd, &rb_prod, &rb_cons, &data);
if (result != EBPF_SUCCESS) {
    // … log error …
    goto Exit;
}

> const volatile uint64_t *prod_offset = &rb_prod->producer_offset; // Producer offset ptr (r only).

Author: Thanks, switched to uint64_t.

bool have_data = producer_offset > consumer_offset;

// Now loop until error.
For(;;) {
  if (!have_data) { // Only wait if we have already caught up.
    // Wait for rb to notify -- or we could spin/poll until *prod_offset > *cons_offset.
    DWORD wait_status = WaitForSingleObject(wait_handle, INFINITE);

    if (wait_status != WAIT_OBJECT_0) { // No notification
      uint32_t wait_err = GetLastError();
      if (wait_err == /* terminal error */) {
        // … log error …
        break;
      }
      // It's possible we still have data, so check even though we got an error.
      producer_offset = ReadAcquire64(prod_offset);
      have_data = producer_offset > consumer_offset;
      if (!have_data) continue;
    } else { // We got notified of new data.
      have_data = true;
    }
  }
  uint64_t remaining = producer_offset - consumer_offset;

  // Check for empty ring.
  if (remaining == 0) {
    have_data = false; // Caught up to producer.
    continue;
  } else if (remaining < EBPF_RINGBUF_HEADER_SIZE) {
    // Bad record or consumer offset out of alignment.
    // … log error …
    break;
  }

>   ebpf_ring_buffer_record_t *record = ebpf_ring_buffer_next_record(data, rb_size, consumer, producer);

Reviewer 6: The function call overhead for going from Go to C is quite high compared to pure Go calls. So relying on these helpers will be bad for performance. Could we document the per-record header instead, maybe even match the Linux layout / flags?

Author: Yes. That is a header-only helper currently in `ebpf_ring_buffer.h`, but I propose we move that to a public header -- then the helper can be used or the header can be parsed in user code.

*More documentation on user-mode API to be added later.*

## Synchronous Consumer Support (proposal)

> The current ringbuffer uses a automatically invoked callbacks to read the ringbuffer.

Reviewer: ```suggestion
The current ringbuffer uses automatically invoked callbacks to read the ringbuffer.
```
grammar

> Linux also supports memory-mapped polling consumers, which aren't currently supported.

Reviewer: Here says "also".  Line 47 says "only".  Which is it?

Author: Clarified, some earlier text shuffling made that unclear.

Linux supports only synchronous consumers, but they can use either synchronous callbacks via libbpf or they can directly access the ring buffer through the mapped memory.

The new API will support 2 consumer types: callbacks and direct access to the mapped producer memory (with an event handle to wait for data).

Asynchronous callback consumer:

> 1. Call `ring_buffer__new` to set up callback with RINGBUF_FLAG_AUTO_CALLBACK specified.

Reviewer: Clarify: is `RINGBUF_FLAG_AUTO_CALLBACK` something internal or exposed to apps?
Is it cross-platform (e.g., in libbpf) or Windows-only?
I don't see it in https://github.com/libbpf/libbpf/blob/master/src/libbpf.h

Author: I added notes to clarify -- on linux there are only synchronous consumers (either using libbpf or directly accessing through mmap), so this is a Windows-specific flag being added for this proposal.

so `RINGBUF_FLAG_AUTO_CALLBACK` should always be specified for asynchronous callbacks for forward-compatibility.

#### Memory mapped consumers

As an alternative to callbacks, Linux ring buffer consumers can directly access the
ring buffer data by calling `mmap()` on a ring_buffer map fd to map the data into user space.
`ring_buffer__epoll_fd()` is used on Linux to get an fd to use with epoll to wait for data.

Windows doesn't have directly compatible APIs to Linux mmap and epoll, so instead we will perfom the mapping
in the eBPF core and use a KEVENT to signal for new data.

For direct memory mapped consumers on Windows, use `ebpf_ring_buffer_get_buffer` to get pointers to the producer and consumer
pages mapped into user space, and `ebpf_ring_buffer_get_wait_handle()` to get the SynchronizationEvent (auto-reset) KEVENT
to use with `WaitForSingleObject`/`WaitForMultipleObject`.

Similar to the linux memory layout, the first page of the producer and consumer memory is the "producer page" and "consumer page",
which contain the 64 bit producer and consumer offsets as the first 8 bytes.

> Only the producer may update the producer offset, and only the consumer may update the consumer offset.

Reviewer: Clarify: is this statement enforced somehow or is it merely a convention?
"may" sounds like it's enforced.  If merely a convention, change both uses of "may" to "should"

};

#define ring_buffer_opts__last_field sz

/**
 * @brief Creates a new ring buffer manager.
 *
 * Only one consumer can be attached at a time, so it should not be called multiple times on an fd.
 *
 * If the return value is NULL the error will be returned in errno.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 * @param[in] sample_cb Pointer to ring buffer notification callback function (if used).
 * @param[in] ctx Pointer to sample_cb callback function context.
 * @param[in] opts Ring buffer options.
 *
 * @returns Pointer to ring buffer manager.
 */
struct ring_buffer *
ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, _Inout_ void *ctx,
		 _In_ const struct ring_buffer_opts *opts);

/**
 * @brief poll ringbuf for new data
 * Poll for available data and consume records, if any are available.
 *
 * If timeout_ms is zero, poll will not wait but only invoke the callback on records that are ready.
 * If timeout_ms is -1, poll will wait until data is ready (no timeout).
 *
 * This function is only supported when automatic callbacks are disabled (see RINGBUF_FLAG_NO_AUTO_CALLBACK).
 *
 * @param[in] rb Pointer to ring buffer manager.

>  * @param[in] timeout_ms maximum time to wait for (in milliseconds).

Reviewer: ```suggestion
 * @param[in] timeout_ms Maximum time to wait for (in milliseconds).
```
nit: consistency

 *

>  * @returns number of records consumed, INT_MAX, or a negative number on error

Reviewer: ```suggestion
 * @returns Number of records consumed, INT_MAX, or a negative number on error
```
nit: For consistency with line 139, 181

 */
int ring_buffer__poll(_In_ struct ring_buffer *rb, int timeout_ms);

/**
 * @brief Frees a ring buffer manager.
 *
 * @param[in] rb Pointer to ring buffer manager to be freed.
 */
void ring_buffer__free(_Frees_ptr_opt_ struct ring_buffer *rb);
```

### New ebpf APIs for mapped memory consumer

```c
/**
 * Get the wait handle to use with WaitForSingleObject/WaitForMultipleObject.
 *
 * Get's the wait handle maintained by the ring buffer manager.
 *
 * Multiple calls will return the same handle, the ring buffer manager will close the handle when destroyed.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 *
 * @returns Wait handle
 */
ebpf_handle ebpf_ring_buffer_get_wait_handle(_In_ struct ring_buffer *rb);

typedef struct _ebpf_ring_buffer_consumer_page
{
    volatile uint64_t consumer_offset; ///< Consumer has read up to this offset.

> } ebpf_ring_buffer_consumer_page_t;

Reviewer: If this is just 64-bits, how is this structure a "page"?

Author: Added comments above structs to clarify.

A whole consumer page and producer page are allocated (as in Linux), but currently only the first part of it is needed. Nothing in the pages outside of these structs should be accessed (for internal/implementation use and future features).

typedef struct _ebpf_ring_buffer_producer_page
{
    volatile uint64_t producer_offset; ///< Producer(s) have reserved up to this offset.
} ebpf_ring_buffer_producer_page_t;

/**

>  * Get pointers to the consumer, producer, and data regions for the mapped ringbuffer memory.

Reviewer: ```suggestion
 * Get pointers to the consumer, producer, and data regions for the mapped ring buffer memory.
```

 *
 * Multiple calls will return the same pointers, as the ring buffer manager only maps the ring once.
 *
 * @param[in] rb Pointer to ring buffer manager.

>  * @param[out] producer pointer* to start of read-only mapped producer pages.

Reviewer: To check: this says "pages".  Does that mean more than one?   Why multiple producer pages and only one consumer page?

Author: Typo - just one producer page and consumer page.

 * @param[out] consumer pointer* to start of read-write mapped consumer page.
 * @param[out] data pointer* to start of read-only double-mapped data pages.
 * @param[out] data_size Size of the mapped data buffer.
 *
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
 */
ebpf_result_t ebpf_ring_buffer_get_buffer(
    _In_ struct ring_buffer *rb,

>     _Out_ ebpf_ring_buffer_consumer_page_t **consumer,
>     _Out_ const ebpf_ring_buffer_producer_page_t **producer,

Reviewer: ```suggestion
    _Out_ ebpf_ring_buffer_consumer_page_t **consumer_page,
    _Out_ const ebpf_ring_buffer_producer_page_t **producer_page,
```
Type and name should match.  This doesn't return a consumer or a producer, if I understand right.

Author: Updated to add `_page` suffix to match names to types.

>     _Out_ uint8_t **data,

Reviewer: Shouldn't this be `_Outptr_result_buffer_(*data_size)` or something similar?

>     _Out_ uint64_t *data_size
>     );

Reviewer: ```suggestion
    _Out_ uint64_t *data_size);
```

/**
 * Map ring buffer memory into user space and get pointers to the consumer, producer, and data regions.
 *
 * Calling this multiple times will map the ring into user-space multiple times.
 *
 * Note: This is a wrapper around ebpf_map_map_buffer.
 *
 * @param[in] map_fd File descriptor to ring buffer map.

>  * @param[out] producer pointer* to start of read-only mapped producer pages.
>  * @param[out] consumer pointer* to start of read-write mapped consumer page.

Reviewer: ```suggestion
 * @param[out] producer Pointer to start of read-only mapped producer pages.
 * @param[out] consumer Pointer to start of read-write mapped consumer page.
```

Author: Fixed this and the other matching doxygen comments.

 * @param[out] data_size Size of the mapped data buffer.
 *
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
 */
ebpf_result_t ebpf_ring_buffer_map_map_buffer(
    fd_t map_fd,
    _Out_ ebpf_ring_buffer_consumer_page_t **consumer,

>     _Out_ const ebpf_ring_buffer_producer_page_t **producer,

Reviewer: See earlier comment about parameter naming.

>     _Out_ const uint8_t **data,

Reviewer: `_Outptr_result_buffer_(*data_size)`?

    _Out_ const uint64_t *data_size
    );

/**
 * Map eBPF map memory into user space.
 *
 * Calling this multiple times will map the memory into user-space multiple times.
 *
 * @param[in] map_fd File descriptor to map.

>  * @param[out] data pointer* to start of mapped memory range.

Reviewer: ```suggestion
 * @param[out] data Pointer to start of mapped memory range.
```

 * @param[out] size Size of the mapped data buffer.
 *
 * @retval EBPF_SUCCESS The operation was successful.
 * @retval other An error occurred.
 */
ebpf_result_t ebpf_map_map_buffer(
    fd_t map_fd,

>     _Out_ const uint8_t **data,

Reviewer: `_Outptr_result_buffer_(*size)`?

>     _Out_ const uint64_t *size
>     );

Reviewer: ```suggestion
    _Out_ const uint64_t *size);
```

/**
 * Set the wait handle that will be signaled for new data.
 *
 * @note Overwrites the wait handle currently stored in the map.
 *
 * @param[in] map_fd File descriptor to ring buffer map.
 *
 * @returns Wait handle
 */
ebpf_result_t ebpf_ring_buffer_map_set_wait_handle(fd_t map_fd, HANDLE handle);
```

> ### Ringbuffer consumer

Reviewer: ```suggestion
### Ring buffer consumer
```

#### mapped memory consumer example

This consumer directly accesses the records from the producer memory and directly updates the consumer offset to show the logic. Normally user code should use the ring buffer helpers
(see second example below) to simplify the logic.

```c++

//
// == Direct mapped memory consumer ==
//

// Open ringbuffer.
fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
if (map_fd == ebpf_fd_invalid) return 1;

// Create wait handle.
HANDLE wait_handle = CreateEvent(nullptr, false, false, nullptr);
if (wait_handle == NULL) {
    // … log error …
    goto Exit;
}

// Set map wait handle.
ebpf_result result = ebpf_ring_buffer_set_wait_handle(map_fd, wait_handle);
if (result != EBPF_SUCCESS) {
    // … log error …
    goto Exit;
}

ebpf_ring_buffer_consumer_page_t *rb_cons; // Read/write consumer page.
const ebpf_ring_buffer_producer_page_t *rb_prod; // Read-only producer page.
const uint8_t* data; // Data region for records.

// Get pointers to the 3 regions.
result = ebpf_ring_buffer_map_map_buffer(map_fd, &rb_prod, &rb_cons, &data);
if (result != EBPF_SUCCESS) {
    // … log error …
    goto Exit;
}

> const volatile uint64_t *prod_offset = &rb_prod->producer_offset; // Producer offset ptr (r only).

Reviewer: ```suggestion
const volatile uint64_t *prod_offset = &rb_prod->producer_offset; // Producer offset ptr (read only).
```

3. Directly read records from the producer pages (and update consumer offset as we read).
4. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.

### Differences from linux API

Linux has only polling ring buffer consumers, even when callbacks are used.
On linux the user code can call `ring_buffer__consume()` to invoke the callback on all available data,
and `ring_buffer__poll()` to wait for data if needed and then consume available data.
Linux consumers can also directly read from the mapped memory by using `mmap()` to map the data
into user space and `ring_buffer__epoll_fd()` to get an epoll wait handle.

On Windows asynchronous events are supported by default,
so nothing extra needs to be done for callbacks to be invoked.

If the `RINGBUF_FLAG_NO_AUTO_CALLBACK` flag is set, callbacks will not automatically be called and `ring_buffer__poll()`
should be called to poll for available data and invoke the callback. On Windows a timeout of zero can be passed to

> `ring_buffer__poll()` to get the same behaviour as `ring_buffer__consume()`.

Reviewer: What is the proposed behavior if `ring_buffer__consume()` is called when the `RINGBUF_FLAG_NO_AUTO_CALLBACK` flag is set?

Author: This proposal only includes `ring_buffer__poll()` (not consume) - That note explains that `ring_buffer__poll(X,0)`  is functionally equivalent to `ring_buffer__consume(X)`
